File: draft-invoices-blocked-by-tax-location/node/stripe-tax-blocked-drafts.mjs
/**
 * Report Stripe draft invoices that cannot finalize because Stripe Tax cannot
 * locate the customer.
 *
 * Read only. One paginated GET and no writes: give this a RESTRICTED key with
 * read access to Invoices. The repair is printed, never performed.
 */
const API = 'https://api.stripe.com/v1';

export const TAX_LOCATION_ERROR = 'customer_tax_location_invalid';
export const TAX_DISABLED_FOR_LOCATION = 'finalization_requires_location_inputs';

// States that mean a human has to touch the customer record or the invoice.
export const ACTIONABLE = ['tax-location', 'tax-dropped', 'needs-address', 'tax-failed'];

/**
 * Classify one draft invoice. Pure, so the rules can be tested without a network.
 * Any of the first three arguments may be null or undefined.
 */
export function verdict(errorCode, taxStatus, disabledReason, autoAdvance) {
  if (errorCode === TAX_LOCATION_ERROR) {
    return ['tax-location',
      'finalization was attempted and refused: Stripe Tax cannot resolve this ' +
      "customer's location"];
  }
  if (disabledReason === TAX_DISABLED_FOR_LOCATION) {
    return ['tax-dropped',
      'Stripe switched automatic tax off so this invoice can finalize; it will ' +
      'be billed and paid with no tax on it'];
  }
  if (taxStatus === 'requires_location_inputs') {
    return ['needs-address',
      'the tax calculation cannot run for want of a location; no finalization ' +
      'attempt has failed yet, but one will'];
  }
  if (taxStatus === 'failed') {
    return ['tax-failed',
      "the calculation failed on Stripe's side; retry the finalization before " +
      'editing the customer'];
  }
  if (errorCode) {
    return ['other-error',
      `finalization is failing for a reason that is not tax: ${errorCode}`];
  }
  if (!autoAdvance) {
    return ['not-advancing',
      'auto_advance is false: this draft is outside the collection workflow ' +
      'rather than blocked by tax'];
  }
  return ['clear', 'no tax finalization problem recorded on this draft'];
}

/** Pull the four fields off an invoice and hand them to verdict(). */
export function classify(inv) {
  const err = inv.last_finalization_error ?? {};
  const tax = inv.automatic_tax ?? {};
  return verdict(err.code, tax.status, tax.disabled_reason, Boolean(inv.auto_advance));
}

async function get(key, path, params = {}) {
  const url = new URL(API + path);
  for (const [k, v] of Object.entries(params)) url.searchParams.set(k, v);
  const res = await fetch(url, { headers: { Authorization: `Bearer ${key}` } });
  if (res.status === 401) {
    throw new Error('401 from Stripe: the key is wrong, or is for the other mode');
  }
  if (!res.ok) throw new Error(`${res.status} from ${url.pathname}`);
  return res.json();
}

export async function drafts(key, limit = 2000) {
  const out = [];
  const params = { status: 'draft', limit: 100 };
  for (;;) {
    const page = await get(key, '/invoices', params);
    const data = page.data ?? [];
    out.push(...data);
    if (data.length === 0 || !page.has_more || out.length >= limit) break;
    params.starting_after = data[data.length - 1].id;
  }
  return out;
}

async function main() {
  const key = (process.env.STRIPE_API_KEY || "dummy-stripe-api-key");
  if (!key) {
    console.error('set STRIPE_API_KEY (use a restricted, read-only key)');
    process.exitCode = 2;
    return;
  }

  let seen = 0;
  const byCustomer = new Map();
  for (const inv of await drafts(key)) {
    seen += 1;
    const [state, detail] = classify(inv);
    if (!ACTIONABLE.includes(state)) continue;
    const cus = (typeof inv.customer === 'object' && inv.customer !== null
      ? inv.customer.id : inv.customer) ?? '<no customer>';
    const entry = byCustomer.get(cus)
      ?? { n: 0, amount: 0, state, detail, first: inv.id };
    entry.n += 1;
    entry.amount += inv.amount_due ?? 0;
    byCustomer.set(cus, entry);
  }

  if (byCustomer.size === 0) {
    console.log(`${'clear'.padEnd(13)} 0 of ${seen} draft invoice(s) blocked on tax location`);
    return;
  }

  const entries = [...byCustomer.entries()].sort((a, b) => b[1].amount - a[1].amount);
  const drafted = entries.reduce((a, [, e]) => a + e.n, 0);
  const atStake = entries.reduce((a, [, e]) => a + e.amount, 0);
  console.warn(`${'tax-blocked'.padEnd(13)} ${entries.length} customer(s), ${drafted} draft(s), ${atStake} in minor units uncollected`);

  for (const [cus, e] of entries.slice(0, 20)) {
    console.warn(`  ${e.state.padEnd(13)} ${cus}  ${e.n} draft(s)  ${e.amount}  ${e.detail}`);
    console.warn(`      GET ${API}/customers/${cus}?expand[]=tax   read tax.automatic_tax and tax.location`);
    console.warn(`      repair: POST ${API}/customers/${cus}  address[country]=..  address[postal_code]=..  tax[validate_location]=immediately`);
    console.warn(`      then: POST ${API}/invoices/${e.first}/finalize`);
  }
  if (entries.length > 20) console.warn(`  ... and ${entries.length - 20} more customer(s)`);
  console.warn('  fix the customer before the invoice; finalizing first either ' +
               'fails again or bills with no tax on it');
  process.exitCode = 1;
}

// Only run when invoked directly. The test file imports this module, and without
// the guard main() would run there too, fail on the missing key, and set a
// non-zero exit code that fails the whole test file even as every test passes.
if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch((err) => { console.error(err.message); process.exitCode = 2; });
}

File: draft-invoices-blocked-by-tax-location/python/stripe_tax_blocked_drafts.py
"""Report Stripe draft invoices that cannot finalize because Stripe Tax cannot
locate the customer.

Read only. One paginated GET and no writes: give this a RESTRICTED key with read
access to Invoices. The repair is printed, never performed, because finalizing an
invoice sends a real bill and disabling tax on one changes what is owed.
"""
import argparse
import logging
import os
import sys

import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("stripe_tax_blocked_drafts")

API = "https://api.stripe.com/v1"

TAX_LOCATION_ERROR = "customer_tax_location_invalid"
TAX_DISABLED_FOR_LOCATION = "finalization_requires_location_inputs"

# States that mean a human has to touch the customer record or the invoice.
ACTIONABLE = ("tax-location", "tax-dropped", "needs-address", "tax-failed")


def verdict(error_code, tax_status, disabled_reason, auto_advance):
    """Classify one draft invoice. Pure, so the rules can be tested without a network.

    `error_code` is last_finalization_error.code, `tax_status` is
    automatic_tax.status, `disabled_reason` is automatic_tax.disabled_reason.
    Any of them may be None. Returns (state, detail).
    """
    if error_code == TAX_LOCATION_ERROR:
        return ("tax-location",
                "finalization was attempted and refused: Stripe Tax cannot "
                "resolve this customer's location")
    if disabled_reason == TAX_DISABLED_FOR_LOCATION:
        return ("tax-dropped",
                "Stripe switched automatic tax off so this invoice can finalize; "
                "it will be billed and paid with no tax on it")
    if tax_status == "requires_location_inputs":
        return ("needs-address",
                "the tax calculation cannot run for want of a location; no "
                "finalization attempt has failed yet, but one will")
    if tax_status == "failed":
        return ("tax-failed",
                "the calculation failed on Stripe's side; retry the finalization "
                "before editing the customer")
    if error_code:
        return ("other-error",
                "finalization is failing for a reason that is not tax: %s"
                % (error_code,))
    if not auto_advance:
        return ("not-advancing",
                "auto_advance is false: this draft is outside the collection "
                "workflow rather than blocked by tax")
    return ("clear", "no tax finalization problem recorded on this draft")


def get(session, path, **params):
    r = session.get(API + path, params=params, timeout=30)
    if r.status_code == 401:
        raise SystemExit("401 from Stripe: the key is wrong, or is for the other mode")
    r.raise_for_status()
    return r.json()


def drafts(session, limit):
    """Page every draft invoice. Newest first, as Stripe sends them."""
    out = []
    params = {"status": "draft", "limit": 100}
    while True:
        page = get(session, "/invoices", **params)
        data = page.get("data", [])
        out.extend(data)
        if not data or not page.get("has_more") or len(out) >= limit:
            break
        params["starting_after"] = data[-1]["id"]
    return out


def classify(inv):
    """Pull the four fields off an invoice and hand them to verdict()."""
    err = inv.get("last_finalization_error") or {}
    tax = inv.get("automatic_tax") or {}
    return verdict(err.get("code"), tax.get("status"), tax.get("disabled_reason"),
                   bool(inv.get("auto_advance")))


def main():
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--top", type=int, default=20,
                    help="how many customers to print")
    ap.add_argument("--max-invoices", type=int, default=2000,
                    help="stop paginating after this many drafts")
    args = ap.parse_args()

    key = os.environ.get("STRIPE_API_KEY")
    if not key:
        log.error("set STRIPE_API_KEY (use a restricted, read-only key)")
        return 2

    s = requests.Session()
    s.headers.update({"Authorization": "Bearer " + key})

    seen = 0
    by_customer = {}
    for inv in drafts(s, args.max_invoices):
        seen += 1
        state, detail = classify(inv)
        if state not in ACTIONABLE:
            continue
        cus = inv.get("customer")
        if isinstance(cus, dict):
            cus = cus.get("id")
        entry = by_customer.setdefault(cus or "<no customer>",
                                       {"n": 0, "amount": 0, "state": state,
                                        "detail": detail, "first": inv.get("id")})
        entry["n"] += 1
        entry["amount"] += inv.get("amount_due") or 0

    if not by_customer:
        log.info("%-13s 0 of %d draft invoice(s) blocked on tax location",
                 "clear", seen)
        return 0

    at_stake = sum(e["amount"] for e in by_customer.values())
    log.warning("%-13s %d customer(s), %d draft(s), %d in minor units uncollected",
                "tax-blocked", len(by_customer),
                sum(e["n"] for e in by_customer.values()), at_stake)

    ranked = sorted(by_customer.items(), key=lambda kv: -kv[1]["amount"])
    for cus, e in ranked[:args.top]:
        log.warning("  %-13s %s  %d draft(s)  %d  %s",
                    e["state"], cus, e["n"], e["amount"], e["detail"])
        log.warning("      GET %s/customers/%s?expand[]=tax   read "
                    "tax.automatic_tax and tax.location", API, cus)
        log.warning("      repair: POST %s/customers/%s  address[country]=..  "
                    "address[postal_code]=..  tax[validate_location]=immediately",
                    API, cus)
        log.warning("      then: POST %s/invoices/%s/finalize", API, e["first"])
    if len(ranked) > args.top:
        log.warning("  ... and %d more customer(s)", len(ranked) - args.top)
    log.warning("  fix the customer before the invoice; finalizing first either "
                "fails again or bills with no tax on it")
    return 1


if __name__ == "__main__":
    sys.exit(main())
